core/src/main/java/org/elasticsearch/index/query/SpanNearQueryBuilder.java

```diff
@@ -19,71 +19,172 @@
 
 package org.elasticsearch.index.query;
 
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.spans.SpanNearQuery;
+import org.apache.lucene.search.spans.SpanQuery;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
 import org.elasticsearch.common.xcontent.XContentBuilder;
 
 import java.io.IOException;
 import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
 
+/**
+ * Matches spans which are near one another. One can specify slop, the maximum number
+ * of intervening unmatched positions, as well as whether matches are required to be in-order.
+ * The span near query maps to Lucene {@link SpanNearQuery}.
+ */
 public class SpanNearQueryBuilder extends AbstractQueryBuilder<SpanNearQueryBuilder> implements SpanQueryBuilder<SpanNearQueryBuilder> {
 
     public static final String NAME = "span_near";
 
-    private ArrayList<SpanQueryBuilder> clauses = new ArrayList<>();
+    static boolean DEFAULT_IN_ORDER = true;
+    static boolean DEFAULT_COLLECT_PAYLOADS = true;
+
+    private final ArrayList<SpanQueryBuilder> clauses = new ArrayList<>();
 
-    private Integer slop = null;
+    private final int slop;
 
-    private Boolean inOrder;
+    private boolean inOrder = DEFAULT_IN_ORDER;
 
-    private Boolean collectPayloads;
+    private boolean collectPayloads = DEFAULT_COLLECT_PAYLOADS;
 
     static final SpanNearQueryBuilder PROTOTYPE = new SpanNearQueryBuilder();
 
+    /**
+     * @param slop controls the maximum number of intervening unmatched positions permitted
+     */
+    public SpanNearQueryBuilder(int slop) {
+        this.slop = slop;
+    }
+
+    /**
+     * only used for prototype
+     */
+    private SpanNearQueryBuilder() {
+        this.slop = 0;
```

Comment:
According to this https://github.com/elastic/elasticsearch/issues/3673 it can officially be "-1", don't know if any other negative values make sense. Lucene doesn't seem to complain. Since this is the prototype setting I can set it to any large negative value probably.